### PR TITLE
fix(salt-user): fix setting grants for the salt user

### PR DESCRIPTION
**Pull request by pprkut**
### PR progress checklist (to be filled in by reviewers)

- [x] Changes to documentation are appropriate (or tick if not required)
- [ ] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

#### Primary type

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [ ] `[ci]`       Changes to the continuous integration configuration
- [ ] `[feat]`     A new feature
- [x] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [ ] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type

- [ ] `[docs]`     Documentation changes
- [ ] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?

No.

### Related issues and/or pull requests

Follow up to #260
More complete fix for #230 

### Describe the changes you're proposing

This is embarrassing :(
I swear I tested the previous pull requests, but I just finished setting up a couple more servers and ran into these issues...

- For some onlyif conditions, the stderr redirect was part of the query sent to mysql, which obviously doesn't work
- The onlyif condition to check grant privileges returns an error string (to stderr) if we still need to execute the state. Redirecting that to stdout yields a false result for the condition and the state to set the grants is never executed. This is fixed by sending stderr to /dev/null. That way stdout response would still be captured, but we don't get false positives from errors.

### Documentation checklist

- [ ] Updated the `README` (e.g. `Available states`).
- [ ] Updated `pillar.example`.

### Testing checklist

- [ ] Included in Kitchen (i.e. under `state_top`).
- [ ] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [ ] Updated the relevant test pillar.

### CODE REVIEW

**Review by javierbertoli — CHANGES_REQUESTED** [review]
@pprkut good catch!

as you're in it, I propose you a couple of changes that I think makes sense and would be great if you apply them.

**Comment by pprkut** [comment]
@javierbertoli Of course!

I can't see them, however :-)

**Comment by javierbertoli** [comment]
> @javierbertoli Of course!
> 
> I can't see them, however :-)

aaargh! It seems that when I clicked to add that comment, didn't save the other text :man_facepalming: 
Re-writing them. Sorry

File: mysql/salt-user.sls [hunk 1]
```diff
@@ -26,8 +26,8 @@ mysql_salt_user_with_salt_user:
     - connection_pass: '{{ mysql_salt_pass }}'
     - connection_charset: utf8
     - onlyif:
-      - mysql --user {{ mysql_salt_user }} --password='{{ mysql_salt_pass|replace("'", "'\"'\"'") }}' -h {{ mysql_host }} --execute="SELECT 1; 2>&1"
-      - VALUE=$(mysql --user {{ mysql_salt_user }} --password='{{ mysql_salt_pass|replace("'", "'\"'\"'") }}' -ss -e "SELECT Grant_priv FROM mysql.user WHERE user = '{{ mysql_salt_user }}' AND host = '{{ host }}';" 2>&1); if [ "$VALUE" = 'Y' ]; then /bin/true; else /bin/false; fi
+      - mysql --user {{ mysql_salt_user }} --password='{{ mysql_salt_pass|replace("'", "'\"'\"'") }}' -h {{ mysql_host }} --execute="SELECT 1;" 2>&1
+      - VALUE=$(mysql --user {{ mysql_salt_user }} --password='{{ mysql_salt_pass|replace("'", "'\"'\"'") }}' -ss -e "SELECT Grant_priv FROM mysql.user WHERE user = '{{ mysql_salt_user }}' AND host = '{{ host }}';" 2>/dev/null); if [ "$VALUE" = 'Y' ]; then /bin/true; else /bin/false; fi
```
Comment (javierbertoli) [inline mysql/salt-user.sls:30]: I think that these two changes might be useful:
1. For the sake of consistency, redirect _stderr_ to `/dev/null` in all the cases
2. For the sake of readability, replace those `replace()`s with [`yaml_dquote`](https://docs.saltproject.io/en/latest/topics/jinja/index.html#yaml-dquote), like
```jinja
- mysql --user {{ mysql_salt_user }} --password={{ mysql_salt_pass | yaml_dquote }} -h {{ mysql_host }} --execute="SELECT 1;" 2>&1
```
which should take care of properly escaping and quoting the password string.

Wdyt?

Comment (pprkut) [reply mysql/salt-user.sls]: Hey! Sorry, this took so long :-(

I made the changes, but I've only been able to verify them on one host so far, and salt crapped out on it so I didn't see the full job output. It worked, but I'd be more comfortable with a bit more testing.

**Review by javierbertoli — APPROVED** [review]
@pprkut They look good to me. Thanks!

**Pull request merged**